src/tallies/filter_particle.cpp

```diff
@@ -0,0 +1,94 @@
+#include "openmc/tallies/filter_particle.h"
+
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+void
+ParticleFilter::from_xml(pugi::xml_node node)
+{
+  auto particles = get_node_array<std::string>(node, "bins");
+
+  // Convert to vector of Particle::Type
+  std::vector<Particle::Type> types;
+  for (auto& p : particles) {
+    if (p == "neutron") {
+      types.push_back(Particle::Type::neutron);
+    } else if (p == "photon") {
+      types.push_back(Particle::Type::photon);
+    } else if (p == "electron") {
+      types.push_back(Particle::Type::electron);
+    } else if (p == "positron") {
+      types.push_back(Particle::Type::positron);
+    }
+  }
+  this->set_particles(types);
+}
+
+void
+ParticleFilter::set_particles(gsl::span<Particle::Type> particles)
+{
+  // Clear existing particles
+  particles_.clear();
+  particles_.reserve(particles.size());
+
+  // Set particles and number of bins
+  for (auto p : particles) {
+    particles_.push_back(p);
+  }
+  n_bins_ = particles_.size();
+}
+
+void
+ParticleFilter::get_all_bins(const Particle* p, int estimator,
+                             FilterMatch& match) const
+{
+  for (auto i = 0; i < particles_.size(); i++) {
+    if (particles_[i] == p->type_) {
+      match.bins_.push_back(i);
+      match.weights_.push_back(1.0);
+    }
+  }
+}
+
+void
+ParticleFilter::to_statepoint(hid_t filter_group) const
+{
+  Filter::to_statepoint(filter_group);
+  std::vector<std::string> particles;
+  for (auto p : particles_) {
+    switch (p) {
+    case Particle::Type::neutron:
```

Comment:
I've seen this code, or at least very similar code, repeated several times so far, here and I believe in another review. I'm guessing from the Particle class declaration's use of an enum for Type that subclasses aren't desired. Instead, perhaps the following two methods in Particle would make sense:
```c++
static string type_to_str(Type type) const {
    switch (type) {
        case Particle::Type::neutron:     return "neutron";
        case Particle::Type::photon:      return "photon";
        case Particle::Type::electron:    return "electron";
        case Particle::Type::positron:    return "positron";
        default:                          return "";
    }
}

static Type str_to_type(string typeName) const {
    if      (typeName == "neutron")       return Particle::Type::neutron:
    else if (typeName == "photon")        return Particle::Type::photon:
    else if (typeName == "electron")      return Particle::Type::electron:
    else if (typeName == "positron")      return Particle::Type::positron:
    else                                  return Particle::Type::UNKNOWN;
}
```